### PR TITLE
[Dubbo-#7100]  Avoid memory leaking when ThreadlessExecutor.waitAnddrain()  throws InterruptionException.

### PR DESCRIPTION
**Pull request by HousonCao**
## What is the purpose of the change

fix issue #7100 

## Brief changelog

1.add try catch on queue.take() to avoid waiting state not be changed when waiting thread has been interrupted.
2. add a unit test in DefaultFutureTest.

## Verifying this change

Link to issue: #7100 .

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-remoting/dubbo-remoting-api/src/test/java/org/apache/dubbo/remoting/exchange/support/DefaultFutureTest.java [hunk 1]
```diff
@@ -116,6 +122,35 @@ public void timeoutSend() throws Exception {
...
+        try {
+            new InterruptThread(Thread.currentThread()).start();
+            executor.waitAndDrain();
+            f.get();
+        } catch (Exception e) {
+            Assertions.assertTrue(e instanceof InterruptedException, "catch exception is not interrupted exception!");
+            System.out.println(e.getMessage());
+        }
+        //waiting timeout check task finishing
+        Thread.sleep(8000);
```
Comment (AlbumenJ) [inline dubbo-remoting/dubbo-remoting-api/src/test/java/org/apache/dubbo/remoting/exchange/support/DefaultFutureTest.java:148]: it would be better if u use shorter timeout

File: dubbo-remoting/dubbo-remoting-api/src/test/java/org/apache/dubbo/remoting/exchange/support/DefaultFutureTest.java [hunk 2]
```diff
@@ -116,6 +122,35 @@ public void timeoutSend() throws Exception {
...
+        // timeout after 5 seconds.
+        Channel channel = new MockedChannel();
+        Request request = new Request(10);
+        ExecutorService sharedExecutor = ExtensionLoader.getExtensionLoader(ExecutorRepository.class)
+                .getDefaultExtension().createExecutorIfAbsent(URL.valueOf("dubbo://127.0.0.1:23456"));
+        ThreadlessExecutor executor = new ThreadlessExecutor(sharedExecutor);
+        DefaultFuture f = DefaultFuture.newFuture(channel, request, 5000, executor);
+        //mark the future is sent
+        DefaultFuture.sent(channel, request);
+        // get operate will throw a timeout exception, because the future is timeout.
```
Comment (HousonCao) [inline dubbo-remoting/dubbo-remoting-api/src/test/java/org/apache/dubbo/remoting/exchange/support/DefaultFutureTest.java:138]: this comment should  change to this: get operate will throw a interrupted exception, because the thread is interrupted.

File: dubbo-remoting/dubbo-remoting-api/src/test/java/org/apache/dubbo/remoting/exchange/support/DefaultFutureTest.java (hunk 1 above)
Comment (HousonCao) [reply dubbo-remoting/dubbo-remoting-api/src/test/java/org/apache/dubbo/remoting/exchange/support/DefaultFutureTest.java:148]: ok，how about  timeout=1500ms ,sleep(2000) and sleep(1000) in InterruptThread

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        with:
+          repository: 'apache/dubbo-samples'
+          ref: master
+      - name: "Download test result"
+        uses: actions/download-artifact@v2
+        with:
+          name: test-result
+          path: test/jobs/
+      - name: "Merge test result"
+        run: ./test/scripts/merge-test-results.sh
```
Comment (chickenlj) [inline .github/workflows/build-and-test.yml:179]: Why is this file changed?

**Review by chickenlj — CHANGES_REQUESTED** [review]

File: dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/ThreadlessExecutor.java [hunk 4]
```diff
@@ -85,8 +85,13 @@ public void waitAndDrain() throws InterruptedException {
...
         }
-
-        Runnable runnable = queue.take();
+        Runnable runnable;
+        try {
+            runnable = queue.take();
+        }catch (InterruptedException e){
+            waiting = false;
+            throw e;
+        }
```
Comment (chickenlj) [inline dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/ThreadlessExecutor.java:94]: This change looks fine to me.

**Comment by chickenlj** [comment]
No response for days, pr closed.

File: .github/workflows/build-and-test.yml (hunk 3 above)
Comment (HousonCao) [reply .github/workflows/build-and-test.yml:179]: this file changed by this commit,when merge branch.
 [Unit Test] Use matrix build workflow & Add Integration Test (#7078)
